api/src/app.ts

```diff
@@ -0,0 +1,193 @@
+import Grant from "./models";
+
+const app = express();
+
+/**
+ * List of available, supported API versions.
+ */
+const apiVersions = [1];
+
+/**
+ * The version in the header is namespaced, specified as:
+ *   gov.usda.fs.nrm.ga.version={VERSION}
+ *
+ * As presently written, versions can only be whole integers.
+ *
+ */
+const apiVersionHeader = /gov\.usda\.fs\.nrm\.ga\.api\.version=(\d+)/;
+
+/**
+ * This matches the first segment of the request path, formatted as:
+ *   v{VERSION}
+ *
+ * As presently written, versions can only be whole integers.
+ *
+ */
+const apiVersionPath = /^v(\d+)$/;
+
+/**
+ * @todo Set using environment variables for security hardening.
+ */
+app.use(function (req, res, next) {
+  res.setHeader("Access-Control-Allow-Origin", "*");
+  res.setHeader("Access-Control-Allow-Methods", "*");
+  res.setHeader("Access-Control-Allow-Headers", "*");
+  next();
+});
+
+const awaitErrorHandlerFactory = (middleware: any) => {
+  return async (request: Request, response: Response, next: NextFunction) => {
+    try {
+      await middleware(request, response, next);
+    } catch (err) {
+      console.error(err);
+      next(err);
+    }
+  };
+};
+
+const versionMessages = {
+  200: {
+    title: "OK",
+    detail: "Everything is fine.",
+  },
+  400: {
+    title: "No API version specified",
+    detail: `An API version must be specified, but no version was specified.
+        To specify a version, use one of the following approaches:
+        (1) Specify the version in the first segment of the path, e.g. \`/v1/grants\`
+        (2) Specify the version in the Accept header: \`Accept: application/vnd.api+json; gov.usda.fs.ga.api.version=1\``,
+  },
+  404: {
+    title: "Version not found",
+    detail: `The version specified is not an available API version. Please choose from the following available versions: ${apiVersions}`,
+  },
+  409: {
+    title: "Conflicting versions requested",
+    detail:
+      "The Accept header and the request path each requested different, conflicting versions. Please ensure that only one version is specified. You can set both the path and header to the same version, or you can request a version using only the path or only the header.",
+  },
+};
+
+/**
+ * Adds a `version` property to Express.Request, so we can attach
+ * that metadata as we resolve the version.
+ */
+declare global {
+  namespace Express {
+    interface Request {
+      version: number;
+    }
+  }
+}
+
+app.use(function (request, response, next) {
+  let [statusCode, version] = resolveVersions(request);
+  if (statusCode == 200) {
+    /**
+     * If the first segment includes the version,
+     * rewrite the path without the version.
+     */
+    if (request.url.match(/^\/v\d+/)) {
+      request.url = `/${request.url.split("/").slice(2).join("/")}`;
+    }
+    request.version = version;
+    next();
+  } else {
+    return response.status(statusCode).json({
+      errors: [
```

Comment:
Because of the API Gateway Lambda proxy integration, it's possible that we might need to pack these JSON responses differently (https://docs.aws.amazon.com/apigateway/latest/developerguide/set-up-lambda-proxy-integrations.html#api-gateway-simple-proxy-for-lambda-output-format). We might think about using a helper method and `NODE_ENV` to change when and how we do that.

Reply:
@neilmb Lambda transforms the incoming response into parameters, e.g. `GET /v1/grants` -> `{ 
"path": "/v1/grants" }`. Is it possible that "the backend Lambda function" referenced is the _wrapper_ around the function code, and that when Express returns a response, Lambda translates it correctly?

Or do we actually have to output that particular response payload manually?

Reply:
From reading the [aws-serverless-express](https://github.com/vendia/serverless-express) docs, it seems like that module does the mapping for us.